htmd/builder/amber.py

```diff
@@ -325,21 +325,37 @@ def build(mol, ff=None, topo=None, param=None, prefix='structure', outdir='./bui
         f.write('loadamberprep {}\n'.format(newname))
     f.write('\n')
 
-    # Detect disulfide bridges if not defined by user
-    if disulfide is None and not ionize:
-        logger.info('Detecting disulfide bonds.')
-        disulfide = detectDisulfideBonds(mol)
-
-    # Fix structure to match the disulfide patching
-    if not ionize and len(disulfide) != 0:
+    # TODO: Remove this once we deprecate the class
+    from htmd.builder.builder import DisulfideBridge
+    from htmd.molecule.molecule import UniqueResidueID
+    if disulfide is not None and len(disulfide) != 0 and isinstance(disulfide[0], DisulfideBridge):
+        newdisu = []
         for d in disulfide:
-            # Rename the residues to CYX if there is a disulfide bond
-            atoms1 = (mol.segid == d.segid1) & (mol.resid == d.resid1)
-            atoms2 = (mol.segid == d.segid2) & (mol.resid == d.resid2)
-            mol.resname[atoms1] = 'CYX'
-            mol.resname[atoms2] = 'CYX'
-            # Remove (eventual) HG hydrogens on these CYS (from proteinPrepare)
-            mol.remove((atoms1 & (mol.name == 'HG')) | (atoms2 & (mol.name == 'HG')), _logger=False)
+            r1 = UniqueResidueID.fromMolecule(mol, 'resid {} and segname {}'.format(d.resid1, d.segid1))
+            r2 = UniqueResidueID.fromMolecule(mol, 'resid {} and segname {}'.format(d.resid2, d.segid2))
+            newdisu.append([r1, r2])
+        disulfide = newdisu
+    # TODO: Remove up to here ----------------------
+
+    if disulfide is not None and len(disulfide) != 0 and isinstance(disulfide[0][0], str):
+        disulfide = convertDisulfide(mol, disulfide)
+
+    # Detect disulfide bridges if not defined by user
+    if not ionize:
+        if disulfide is None:
+            logger.info('Detecting disulfide bonds.')
+            disulfide = detectDisulfideBonds(mol)
+
+        # Fix structure to match the disulfide patching
+        if len(disulfide) != 0:
+            for d in disulfide:
+                # Rename the residues to CYX if there is a disulfide bond
+                atoms1 = d[0].selectAtoms(mol, ignore='resname', indexes=False)
```

Comment:
Why the `ignore='resname'`? As long as it is consistent with the actual adding of the disulfide bonds in line 391-2, I guess it's OK, but I don't understand your reasoning. Is it because of the double building? If so, I think it's better to pass all the disulfide detection into the `if not ionize`

Reply:
Yes because CYS gets renamed to CYX to create the bridge. Hmmm could we do the disu bonding always post ionize? I doesn't affect the total charge right?

Reply:
In the current code (before your changes), it seems like all disulfide machinery is inside the `if not ionize`. It does not affect charge: cystein is either protonated (charge 0) or bound to another cystein (charge 0). Unless the user is working with a deprotonated cystein, but that's an advanced user which should not let disulfides to be defined in an automated way, I guess?

Reply:
Ok, fixed now. All disu stuff is post ionize and no need for ignoring the resname